identity/aziot-dps-client-async/src/lib.rs

```diff
@@ -13,323 +13,339 @@
+                return Err(std::io::Error::new(std::io::ErrorKind::Other, res.message));
+            }
+
+            _ => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "malformed HTTP response",
+                ))
+            }
+        };
 
-        let mut device_id_certs = {
-            let device_id_certs = cert_client
-                .get_cert(&identity_cert)
+        Ok(res)
+    }
+
+    async fn get_sas_connector(
+        &self,
+        registration_id: &str,
+        key_handle: &str,
+    ) -> Result<
+        (
+            hyper_openssl::HttpsConnector<hyper::client::HttpConnector>,
+            String,
+        ),
+        std::io::Error,
+    > {
+        let key_handle = self.key_client.load_key(key_handle).await?;
+
+        let token = {
+            let expiry = chrono::Utc::now()
+                + chrono::Duration::from_std(std::time::Duration::from_secs(30))
+                    .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+            let expiry = expiry.timestamp().to_string();
+            let audience = format!("{}/registrations/{}", self.scope_id, registration_id);
+
+            let resource_uri = percent_encoding::percent_encode(
+                audience.to_lowercase().as_bytes(),
+                DPS_ENCODE_SET,
+            )
+            .to_string();
+            let sig_data = format!("{}\n{}", &resource_uri, expiry);
+
+            let signature = self
+                .key_client
+                .sign(
+                    &key_handle,
+                    aziot_key_common::SignMechanism::HmacSha256,
+                    sig_data.as_bytes(),
+                )
                 .await
                 .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
-            let device_id_certs = openssl::x509::X509::stack_from_pem(&device_id_certs)
-                .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?
-                .into_iter();
-            device_id_certs
+
+            let signature = base64::encode(&signature);
+
+            let token = url::form_urlencoded::Serializer::new(format!("sr={}", resource_uri))
+                .append_pair("sig", &signature)
+                .append_pair("se", &expiry)
+                .finish();
+            token
         };
-        let client_cert = device_id_certs.next().ok_or_else(|| {
-            std::io::Error::new(std::io::ErrorKind::Other, "device identity cert not found")
-        })?;
-        tls_connector
-            .set_certificate(&client_cert)
+
+        let token = format!("SharedAccessSignature {}", token);
+
+        let tls_connector = hyper_openssl::HttpsConnector::new()
             .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
-        for cert in device_id_certs {
+        Ok((tls_connector, token))
+    }
+
+    async fn get_x509_connector(
```

Comment:
The `get_{sas,x509}_connector` fns are duplicated in `aziot-hub-client-async`.

Perhaps leave them as free functions, put them in a common crate and have both `aziot-{dps,hub}-client-async` use them.

Reply:
Oh, hey, would you look at that. Neato.
I'm very for reducing duplication when possible, so I'll go ahead and break those out into their own crate. Any suggestion on the crate name? Off the top of my head... `aziot-auth-client-async-common`? A bit of a mouthful, but it's certainly descriptive...

Reply:
`aziot-http-client(-async)(-common)` perhaps. I'd be fine with just `aziot-http-client`

`-async` is optional since there's no `-sync`. (`aziot-cert-client-async` is a counter-point, but that is symmetric with `aziot-key-client-async`. Though I'm not opposed to renaming it to `aziot-cert-client`)

Reply:
While I'm all for removing the `-async` suffxes across the repo, that's best left for a separate PR.

As for naming... what about `aziot-cloud-client-async`? `cloud` encapsulates both the DPS and the Hub, while leaving it open for additional shared code in the future (as opposed to just `auth` stuff)

Reply:
>While I'm all for removing the `-async` suffxes across the repo, that's best left for a separate PR.

I didn't say anything about renaming any existing crates in this PR. I'm talking about your new crate, for which `-async` is optional.

>As for naming... what about `aziot-cloud-client-async-common` ? `cloud` encapsulates both the DPS and the Hub, while leaving it open for additional shared code in the future (as opposed to just `auth` stuff)

It's not DPS- or IoT Hub-specific.

Reply:
Sorry, let me rephrase what I meant - I'll leave `-async` in the name to match the other clients in the repo. At some point, it would be nice to normalize all the crate names to be async-first, and renaming the few synchronous clients with the `-sync` suffix.

* * *

Sure, but this crate contains shared code for talking to the cloud, so it seems somewhat appropriate (even in the concrete contents of the crate isn't strictly cloud-specific at the moment).

I don't have super strong feelings about the name, so let me know if this works / what I should rename it to, and I'll go ahead and push an update.